core/jsonvalue.go

```diff
@@ -0,0 +1,137 @@
+package core
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/vektah/gqlparser/v2/ast"
+)
+
+func (*JSONValue) Type() *ast.Type {
+	return &ast.Type{
+		NamedType: "JSONValue",
+		NonNull:   true,
+	}
+}
+
+type JSONValue struct {
+	data []byte
+}
+
+// NewJSONValue constructs a JSONValue from any Go value.
+func NewJSONValue(v any) (*JSONValue, error) {
+	b, err := json.Marshal(v)
+	if err != nil {
+		return &JSONValue{}, err
+	}
+	return &JSONValue{data: b}, nil
+}
+
+// Set sets value at path (dot-separated).
+// "" or "." replaces the whole value.
+func (o *JSONValue) Set(path string, value any) (*JSONValue, error) {
```

Comment:
WDYT about supporting arbitrary JSONPath? There's a go implementation we could use here: https://github.com/oliveagle/jsonpath